win_iis_webapplication: add authentication parameters

mhunsber opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
When setting up an application in IIS, you can specify the authentication type as either Application User (pass-through authentication) or specific user with a username and password. This just adds the parameters to support that.

I also added tests since it didn't seem like this module had any integration tests yet.

Added the following parameters to win_iis_webapplication: connect_as, username, password.

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Feature Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
win_iis_webapplication

##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
you can do something like this
```paste below
win_iis_webapplication:
  state: present
  site: ACME
  name: api
  connect_as: specific_user
  username: acme_user
  password: secret
```

Commit:
add connect_as, username, password parameters
add tests

Commit:
fixed reference to undefined variable.
added version added to new options.

ShachafGoldstein commented on lib/ansible/modules/windows/win_iis_webapplication.ps1:
```diff
@@ -13,6 +13,9 @@ $site = Get-AnsibleParam -obj $params -name "site" -type "str" -failifempty $tru
 $state = Get-AnsibleParam -obj $params -name "state" -type "str" -default "present" -validateset "absent","present"
 $physical_path = Get-AnsibleParam -obj $params -name "physical_path" -type "str" -aliases "path"
 $application_pool = Get-AnsibleParam -obj $params -name "application_pool" -type "str"
+$connect_as = Get-AnsibleParam -obj $params -name 'connect_as' -type 'str' -validateset 'specific_user', 'pass_through'
```
You should add a default option of pass_through

mhunsber replied:
If this were a new module, I would make that the default option, but there could be tasks out there using this module where they have changed the authentication to specific user a different way. Making a default value on this option would break that configuration.
Setting a default also wouldn't be consistent with the behavior of application_pool and physical_path, where they are only set if the user passes them as parameters.

ShachafGoldstein replied:
you are absolutely right, missed that

ShachafGoldstein commented:
Should add a changelog fragment and update the .py file for documentation

Commit:
add changelog fragment

Commit:
fix line endings

mhunsber commented:
@ShachafGoldstein I added the changelog fragment. The .py documentation was already updated to show the new options.
I don't really know what happened during the integration test. Looks like it failed when trying to [backup the iis configuration](https://github.com/ansible/ansible/pull/56033/files#diff-e804605f0640a24d3321167d3b4e5febR23).

ShachafGoldstein commented:
close and open the PR to make it run the tests again

mhunsber commented:
Failed on 2012-R2 this time, same error message as before.

ShachafGoldstein commented on test/integration/targets/win_iis_webapplication/tasks/main.yml:
```diff
@@ -0,0 +1,45 @@
+# Cannot use win_feature to install IIS on Server 2008.
+# Run a brief check and skip hosts that don't support
+# that operation
+- name: check if win_feature will work on test host
```
Why aren't you simply checking the Powershell / Os version using facts?

mhunsber replied:
I had stolen this from the win_iis_webapppool tests. Since we're running into issues I'll definitely rework it to use facts.

mhunsber replied:
using ansible facts now

ShachafGoldstein commented on test/integration/targets/win_iis_webapplication/tasks/main.yml:
```diff
@@ -0,0 +1,45 @@
+      dest: '{{ remote_tmp_dir }}\applicationHost.config'
+      remote_src: yes
+
+  - name: set version of IIS for tests
```
Never used

mhunsber replied:
removed it.

ShachafGoldstein commented:
Only thing I can think of is that the copy somehow damages the file.
try coping to a different path or skip the copy just to check it maybe?

Commit:
use ansible facts to determine os version
remove unused iis version check
test checksum of iis configuration after backup

Commit:
correct assertion

Commit:
added more cleanup tasks.

mhunsber commented:
Adding some extra cleanup tasks before restoring the previous IIS configuration seemed to work. After looking closer I realized it was failing on the restore copy because IIS was still using the configuration file. Then the tests were run again in verbose mode, which is where we were seeing the failure in the website module.

Commit:
version added is now 2.10

Commit:
skip server 2008 r2 for now

Commit:
run tests on server 2012 and higher

mhunsber commented:
No idea why Server 2008R2 doesn't like when application_pool is null, I'll just skip it for now and run the tests on 2012 and higher. I'm guessing that it is something to do with the PowerShell version or web administration module version on the host. Maybe someone in the future will decide to make it work on 2008R2.

jborean93 commented:
Thanks for the contribution and adding the tests as well. Apologies for the delay and thanks for sticking with it.
